Schema updates for current and upcoming asdf changes

braingram opened the pull request:
## Changes

asdf is working on making the current development version of the ASDF standard (1.6.0) stable (see https://github.com/asdf-format/asdf/pull/1744 for details). As the 1.6.0 version contains new versions of a few core schemas (including [ndarray](https://github.com/asdf-format/asdf-standard/blob/main/resources/schemas/stsci.edu/asdf/core/ndarray-1.1.0.yaml)). This PR updates the weldx schemas for these upcoming changes (and applies a few other related fixes).

- update schema examples that use `ndarray-1.0.0` adding a `asdf-standard` requirement to force these examples to use ASDF standard 1.5.0 (where `ndarray-1.0.0` exists)
- update schema "id"s to match "uri"s used to register the schemas with asdf
- update manifest generation script to use the uri prefix used to register the manifests (see below)

For the manifests, the `id` in the files follow the form `.../weldx/manifests/...`
https://github.com/BAMWelDX/weldx/blob/2bceb68405385229190cb00bb71dec95e941bed9/weldx/manifests/weldx-0.1.2.yaml#L3
However these files are registered:
https://github.com/BAMWelDX/weldx/blob/2bceb68405385229190cb00bb71dec95e941bed9/weldx/asdf/extension.py#L27-L33
using the `WELDX_EXTENSION_URI_BASE` uri prefix:
https://github.com/BAMWelDX/weldx/blob/2bceb68405385229190cb00bb71dec95e941bed9/weldx/asdf/constants.py#L9
which uses `.../weldx/extensions`. This means that these manifest resources are available (via the asdf `resource_manager`) using `../weldx/extensions/...` uris yet have ids with `.../weldx/manifests/`

```python
>> import asdf
>> manifest = asdf.schema.load_schema("asdf://weldx.bam.de/weldx/extensions/weldx-0.1.2")
>> manifest['id']
asdf://weldx.bam.de/weldx/manifests/weldx-0.1.2
```
This PR changes the `id`s to match the `uri`s.

This PR also updates `SerializationContext` imports to use `asdf.extension` for asdf version >= 3.0.0 to avoid `DeprecationWarning` messages form importing `asdf.asdf` and `SerializationContext` from `asdf.asdf`.

## Related Issues

Fixes #900 

## Checks

- [ ] updated `CHANGELOG.md`
- [ ] updated `tests/`
- [ ] updated `doc/`
- [ ] update example/tutorial notebooks
- [ ] update manifest file

CagtayFabry commented:
Hi @braingram,
thank you for the explanations.

I am planning a rework on some of the code for asdf>=3 (hopefully soon)
In that light would it make sense to include this as a transition and remove any asdf 2.x related code in the next major release of weldx?

braingram commented:
Thanks for taking a look!

> I am planning a rework on some of the code for asdf>=3 (hopefully soon) In that light would it make sense to include this as a transition and remove any asdf 2.x related code in the next major release of weldx?

That sounds great. Let me know if there's anything I can do to help (and feel free to ping me when the changes are ready for review or if you run into any issues). There are a few changes in 3.1.0 (changelog [here](https://github.com/asdf-format/asdf/releases/tag/3.1.0)) that might be useful including:
- starting the transition from `copy_arrays` to `memmap` (in preparation for making `memmap=False` as the default) See: https://github.com/asdf-format/asdf/pull/1667
- introduction of `asdf.util.load_yaml` for loading just the yaml portion of an asdf file (as python builtin types or asdf `tagged` objects): https://github.com/asdf-format/asdf/pull/1700

CagtayFabry commented:
Thank you, I followed the 3.0/3.1 release notes and the changes will make for some big QOL improvements here!

braingram commented:
FYI the asdf development branch will soon be transitioning to asdf 4.0 where ASDF standard 1.6.0 will be the default.

braingram commented:
Following up on https://github.com/BAMWelDX/weldx/pull/952 I modified this PR to also unpin asdf.

I also noticed that there is an import of `AsdfSpec` that's used in `format_tag`:
https://github.com/BAMWelDX/weldx/blob/c8457bf3d38e37a2f53c60b37ecf23a01f4509e6/weldx/asdf/types.py#L120-L121
`AsdfSpec` was deprecated in asdf 3.3.0 and I can't find any places in weldx where an `AsdfSpec` is provided to `format_tag` so I updated this PR to remove the `AsdfSpec` usage. See https://github.com/asdf-format/asdf/issues/1772 for motivation why `AsdfSpec` was removed.

braingram commented:
Also tests like this one:
https://github.com/BAMWelDX/weldx/blob/c8457bf3d38e37a2f53c60b37ecf23a01f4509e6/weldx/tests/asdf_tests/test_weldx_file.py#L324-L325
appear to be relying on asdf validating a tree provided via `AsdfFile(tree)`. This behavior was deprecated in asdf 3.1.0 (see https://github.com/asdf-format/asdf/pull/1691) and removed in asdf 4.0.0.

I modified this PR to call `AsdfFile.validate` during `WeldxFile._write_tree` for asdf >= 4.0.0. (asdf 4.0.0 validates on write so the modification here might be a bit confusing. I picked this function since it's what's called during [WeldxFile.__init__](https://github.com/BAMWelDX/weldx/blob/c8457bf3d38e37a2f53c60b37ecf23a01f4509e6/weldx/asdf/file.py#L280)).
